zkvm/src/contract.rs

```diff
@@ -52,27 +53,17 @@ enum FrozenItem {
 /// Note: values do not necessarily have open commitments. Some can be reblinded,
 /// others can be passed-through to an output without going through `cloak` and the constraint system.
 #[derive(Clone, Debug)]
-struct FrozenValue {
+pub struct FrozenValue {
     qty: Commitment,
     flv: Commitment,
 }
 
 impl Input {
-    /// Creates a "frozen contract" from payload (which is a vector of (qty, flv)) and pred.
-    /// Serializes the contract, and uses the serialized contract and txid to generate a utxo.
+    /// Serializes the contract, and uses the serialized contract and txid
+    /// to generate a utxo.
     /// Returns an Input with the contract, txid, and utxo.
-    pub fn new<I>(payload: I, predicate: Predicate, txid: TxID) -> Self
-    where
-        I: IntoIterator<Item = (Commitment, Commitment)>,
-    {
-        let payload: Vec<FrozenItem> = payload
-            .into_iter()
-            .map(|(qty, flv)| FrozenItem::Value(FrozenValue { qty, flv }))
-            .collect();
-
-        let prev_output = Output { payload, predicate };
+    pub fn new(prev_output: Output, txid: TxID) -> Self {
         let utxo = UTXO::from_output(&prev_output.clone().to_bytes(), &txid);
```

Comment:
This output.clone.to_bytes() is unfortunate. There's no win in consuming `self` here because we need to allocate a Vec anyway. Let's change it to take `&self` and replace `.clone().to_bytes()` with `.to_bytes()` here and in the other place below.